msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs

```diff
@@ -0,0 +1,202 @@
+
+            AuthenticationRequestParameters parameters = new AuthenticationRequestParameters()
+            {
+                Authority = authority,
+                // todo: what is this? SliceParameters = "key1=value1%20with%20encoded%20space&key2=value2",
+                ClientId = TestConstants.ClientId,
+                Scope = TestConstants.Scope,
+                TokenCache = _cache,
+                RequestContext = new RequestContext(new MsalLogger(Guid.NewGuid(), null))
+            };
+
+            RequestTestsCommon.MockInstanceDiscoveryAndOpenIdRequest();
+
+            var expectedScopes = new SortedSet<string>();
+            expectedScopes.UnionWith(TestConstants.Scope);
+            expectedScopes.Add("openid");
+            expectedScopes.Add("offline_access");
+            expectedScopes.Add("profile");
+
+            // Mock Handler for device code request
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Post,
+                PostData = new Dictionary<string, string>()
+                {
+                    {OAuth2Parameter.ClientId, TestConstants.ClientId},
+                    {OAuth2Parameter.Scope, expectedScopes.AsSingleString()}
+                },
+                ResponseMessage = CreateDeviceCodeResponseSuccessMessage()
+            });
+
+            // Mock Handler for devicecode->token exchange request
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Post,
+                PostData = new Dictionary<string, string>()
+                {
+                    {OAuth2Parameter.ClientId, TestConstants.ClientId},
+                    {OAuth2Parameter.Scope, expectedScopes.AsSingleString()}
+                },
+                ResponseMessage = MockHelpers.CreateSuccessTokenResponseMessage()
+            });
+
+            DeviceCodeResult actualDeviceCodeResult = null;
+            DeviceCodeRequest request = new DeviceCodeRequest(parameters, result => { actualDeviceCodeResult = result; });
+            var task = request.RunAsync();
+            task.Wait();
+            var authenticationResult = task.Result;
+            Assert.IsNotNull(authenticationResult);
+            Assert.IsNotNull(actualDeviceCodeResult);
+
+            Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId);
+            Assert.AreEqual("BAQABAAEAAADXzZ3ifr-GRbDT45zNSEFEfU4P-bZYS1vkvv8xiXdb1_zX2xAcdcfEoei1o-t9-zTB9sWyTcddFEWahP1FJJJ_YVA1zvPM2sV56d_8O5G23ti5uu0nRbIsniczabYYEr-2ZsbgRO62oZjKlB1zF3EkuORg2QhMOjtsk-KP0aw8_iAA", actualDeviceCodeResult.DeviceCode);
+            Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId);
+
+            // todo: figure out how to unit test this one since it's calculated from current time based on the datetime offset retrieved in the response
+            // Assert.AreEqual(5, actualDeviceCodeResult.ExpiresOn); 

```

Comment:
There are a series of tests in DeviceProfileTests that check the timeout exceptions. But, as stated above, we if we centralize this logic in core, we might be able to reuse those tests.